mono/mini/tiered.c

```diff
@@ -0,0 +1,137 @@
+
+					for (; patchsites != NULL; patchsites = patchsites->next) {
+						gpointer patchsite = (gpointer) patchsites->data;
+
+						mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_TIERED, "tiered: patching %p with patch_kind=%s @ tier_level=%d", patchsite, patch_kind_str [patch_kind], tier_level);
+						mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_TIERED, "\t-> caller=%s", mono_pmip (patchsite));
+						mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_TIERED, "\t-> callee=%s", mono_method_full_name (ppc->target_method, TRUE));
+
+						gboolean success = patchers [patch_kind] (ppc, patchsite);
+
+						if (!success)
+							mono_trace (G_LOG_LEVEL_WARNING, MONO_TRACE_TIERED, "tiered: couldn't patch %p with target %s, dropping it.", patchsite, mono_method_full_name (ppc->target_method, TRUE));
+					}
+					g_hash_table_remove (callsites_hash [patch_kind], ppc->target_method);
+					g_slist_free (patchsites);
+				}
+				g_free (ppc);
+			}
+			g_slist_free (ppcs);
+		}
+		mono_coop_mutex_unlock (&compilation_mutex);
+	}
+}
+
+void
+mini_tiered_init (void)
+{
+	ERROR_DECL (error);
+
+	mono_counters_init ();
+	mono_counters_register ("Methods promoted", MONO_COUNTER_TIERED | MONO_COUNTER_LONG, &mini_tiered_stats.methods_promoted);
+
+	mono_coop_cond_init (&compilation_wait);
+	mono_coop_mutex_init (&compilation_mutex);
+
+	mono_thread_create_internal (mono_domain_get (), compiler_thread, NULL, MONO_THREAD_CREATE_FLAGS_THREADPOOL, error);
+	mono_error_assert_ok (error);
+}
+
+void
+mini_tiered_register_callsite_patcher (CallsitePatcher func, int level)
+{
+	g_assert (level < NUM_TIERS);
+
+	patchers [level] = func;
+}
+
+void
+mini_tiered_record_callsite (gpointer ip, MonoMethod *target_method, int patch_kind)
+{
+	if (!callsites_hash [patch_kind])
+		callsites_hash [patch_kind] = g_hash_table_new (NULL, NULL);
+
+	GSList *patchsites = g_hash_table_lookup (callsites_hash [patch_kind], target_method);
+	patchsites = g_slist_prepend (patchsites, ip);
+	g_hash_table_insert (callsites_hash [patch_kind], target_method, patchsites);
+}
+
+void
+mini_tiered_inc (MonoDomain *domain, MonoMethod *method, MiniTieredCounter *tcnt, int tier_level)
+{
+	if (G_UNLIKELY (tcnt->hotness == threshold [tier_level] && !tcnt->promoted)) {
+		tcnt->promoted = TRUE;
+		mini_tiered_stats.methods_promoted++;
+
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_TIERED, "tiered: queued %s", mono_method_full_name (method, TRUE));
+
+		MiniTieredPatchPointContext *ppc = g_new0 (MiniTieredPatchPointContext, 1);
+		ppc->domain = domain;
+		ppc->target_method = method;
+		ppc->tier_level = tier_level;
+
+		mono_coop_mutex_lock (&compilation_mutex);
+		compilation_queue [tier_level] = g_slist_append (compilation_queue [tier_level], ppc);
+		mono_coop_mutex_unlock (&compilation_mutex);
+		mono_coop_cond_signal (&compilation_wait);
+	} else if (!tcnt->promoted) {
+		/* FIXME: inline that into caller */
+		tcnt->hotness++;
+	}
+}
+
+#endif
```

Comment:
```suggestion
#else
MONO_EMPTY_SOURCE_FILE(tiered)
#endif
```

And then unconditionally include the source file.  That will make it a bit easier to use on Windows with our msvc projects.

Reply:
There is also similar pattern, no macro and unconditional:
```
C:\s\mono2\mono\metadata\mono-route.c(160):extern const char mono_route_empty_file_no_warning;
C:\s\mono2\mono\metadata\mono-route.c(161):const char mono_route_empty_file_no_warning = 0;
```

Reply:
@lambdageek Thanks, I didn't know about it!